File: app/src/main/java/com/originb/inkwisenote/modules/smartnotes/ui/SmartNotebookAdapter.java
package com.originb.inkwisenote.modules.smartnotes.ui;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.activity.ComponentActivity;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.originb.inkwisenote.common.Logger;
import com.originb.inkwisenote.R;
import com.originb.inkwisenote.modules.smartnotes.data.AtomicNoteEntity;
import com.originb.inkwisenote.modules.backgroundjobs.BackgroundOps;
import com.originb.inkwisenote.modules.handwrittennotes.ui.HandwrittenNoteHolder;
import com.originb.inkwisenote.modules.repositories.Repositories;
import com.originb.inkwisenote.modules.repositories.SmartNotebook;
import com.originb.inkwisenote.modules.repositories.SmartNotebookRepository;
import com.originb.inkwisenote.modules.smartnotes.data.NoteType;
import com.originb.inkwisenote.modules.textnote.TextNoteHolder;
import lombok.Setter;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

public class SmartNotebookAdapter extends RecyclerView.Adapter<NoteHolder> {

    private Logger logger = new Logger("SmartNotebookAdapter");

    private final ComponentActivity parentActivity;

    @Setter
    private SmartNotebook smartNotebook;
    private final SmartNotebookRepository smartNotebookRepository;

    // noteId to card mapping
    private final Map<Long, NoteHolder> noteCards = new HashMap<>();

    public SmartNotebookAdapter(ComponentActivity parentActivity,
                                SmartNotebook smartNotebook) {
        this.parentActivity = parentActivity;
        this.smartNotebook = smartNotebook;
        this.smartNotebookRepository = Repositories.getInstance().getSmartNotebookRepository();
    }

    @NonNull
    @NotNull
    @Override
    public NoteHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int position) {
        AtomicNoteEntity atomicNote = smartNotebook.getAtomicNotes().get(position);
        View itemView;

        if (NoteType.HANDWRITTEN_PNG.equals(atomicNote.getNoteType())) {
            itemView = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.note_drawing_layout, parent, false);

            return new HandwrittenNoteHolder(itemView, parentActivity, smartNotebookRepository);
        } else if (NoteType.TEXT_NOTE.equals(atomicNote.getNoteType())) {
            itemView = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.note_text_layout, parent, false);
            return new TextNoteHolder(itemView, parentActivity, smartNotebookRepository);
        }

        itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.note_init_layout, parent, false);

        return new InitNoteHolder(itemView, parentActivity, smartNotebookRepository, this);
    }

    @Override
    public void onBindViewHolder(@NonNull @NotNull NoteHolder noteHolder, int position) {
        AtomicNoteEntity atomicNote = smartNotebook.getAtomicNotes().get(position);
        noteHolder.setNote(smartNotebook.getSmartBook().getBookId(), atomicNote);
        noteCards.put(atomicNote.getNoteId(), noteHolder);
    }

    @Override
    public int getItemViewType(int position) {
        return position;
    }

    @Override
    public int getItemCount() {
        return smartNotebook.atomicNotes.size();
    }

    public void saveNote(String noteTitle) {
        if (smartNotebook == null) return;
        BackgroundOps.execute(() -> {
            for (NoteHolder noteHolder : noteCards.values()) {
                noteHolder.saveNote();
            }

            // update title
            smartNotebook.getSmartBook().setTitle(noteTitle);
            smartNotebookRepository.updateNotebook(smartNotebook);
        });
    }

    public void removeNoteCard(long noteId) {
        int position = noteCards.get(noteId).getAdapterPosition();
        noteCards.remove(noteId);
        notifyItemRemoved(position);
    }

    // this function assumes that either the smartNotebook has updated
// notes and pages or
// all new notes or pages are inserted after current index so that
// the note and page at this index is not affected.
    public void saveNotebookPageAt(int currentVisibleItemIndex, AtomicNoteEntity atomicNote) {
        if (!noteCards.containsKey(atomicNote.getNoteId())) {
            return;
        }

        NoteHolder noteHolder = noteCards.get(atomicNote.getNoteId());
        BackgroundOps.execute(noteHolder::saveNote);
    }


}

File: androidApp/src/main/java/com/originb/inkwisenote2/modules/smartnotes/ui/SmartNotebookAdapter.kt
package com.originb.inkwisenote2.modules.smartnotes.ui

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import com.originb.inkwisenote2.R
import com.originb.inkwisenote2.common.Logger
import com.originb.inkwisenote2.modules.backgroundjobs.BackgroundOps.Companion.execute
import com.originb.inkwisenote2.modules.handwrittennotes.data.HandwrittenNoteRepository
import com.originb.inkwisenote2.modules.ocr.data.NoteOcrTextsDao
import com.originb.inkwisenote2.modules.repositories.SmartNotebook
import com.originb.inkwisenote2.modules.repositories.SmartNotebookRepository
import com.originb.inkwisenote2.modules.smartnotes.data.AtomicNoteEntity
import com.originb.inkwisenote2.modules.smartnotes.data.NoteHolderData
import com.originb.inkwisenote2.modules.textnote.data.TextNotesDao

class SmartNotebookAdapter(
    private val parentActivity: AppCompatActivity,
    private var smartNotebook: SmartNotebook,
    private val smartNotebookRepository: SmartNotebookRepository,
    private val handwrittenNoteRepository: HandwrittenNoteRepository?,
    private val textNotesDao: TextNotesDao?,
    private val noteOcrTextDao: NoteOcrTextsDao?
) : RecyclerView.Adapter<FragmentViewHolder>() {
    // noteId to card mapping
    private val noteCards: MutableMap<Long?, FragmentViewHolder?> = HashMap<Long?, FragmentViewHolder?>()

    fun setSmartNotebook(smartNotebook: SmartNotebook) {
        this.smartNotebook = smartNotebook
        notifyDataSetChanged()
    }

    fun refreshFragments() {
        notifyDataSetChanged()
    }

    fun setSmartNotebook(smartNotebook: SmartNotebook, indexOfUpdatedNote: Int) {
        this.smartNotebook = smartNotebook
        notifyItemInserted(indexOfUpdatedNote)
    }

    override fun getItemViewType(position: Int): Int {
        return position
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FragmentViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.fragment_note_page, parent, false)

        // Create a truly unique ID for the fragment container
        val holder = FragmentViewHolder(
            this, view, parentActivity,
            handwrittenNoteRepository, textNotesDao,
            smartNotebookRepository, noteOcrTextDao
        )
        holder.fragmentContainer.setId(viewType + 1)

        return holder
    }

    override fun onBindViewHolder(holder: FragmentViewHolder, position: Int) {
        val atomicNotes: MutableList<AtomicNoteEntity> = smartNotebook.atomicNotes

        if (position < 0 || position >= atomicNotes.size) return

        val atomicNote = atomicNotes[position]
        holder.setNote(smartNotebook, atomicNotes[position], position)
        noteCards[atomicNote.noteId] = holder
    }

    fun updateNoteType(atomicNote: AtomicNoteEntity, newNoteType: String?) {

        val position: Int = smartNotebook.atomicNotes.indexOf(atomicNote)
        if (position == -1) {
            return
        }
        atomicNote.noteType = newNoteType
        execute { smartNotebookRepository.updateNotebook(smartNotebook, parentActivity) }
        notifyItemChanged(position)
    }

    fun removeNoteCard(noteId: Long) {
        if (!noteCards.containsKey(noteId)) return

        val position = noteCards[noteId]!!.getAdapterPosition()
        noteCards.remove(noteId)
        notifyItemRemoved(position)
    }

    override fun getItemCount(): Int {
        return smartNotebook.atomicNotes.size
    }

    fun getNoteData(noteId: Long): NoteHolderData? {
        return noteCards[noteId]!!.noteHolderData
    }

    fun setNoteData(index: Int, currentNote: AtomicNoteEntity) {
        if (noteCards.containsKey(currentNote.noteId)) {
            noteCards[currentNote.noteId]!!.setNote(smartNotebook, currentNote, index)
        }
    }

    companion object {
        private val logger = Logger("SmartNotebookAdapter")
    }
}




































